### DISCUSSION TITLE
How to port fsharp on new CPU architecture ?

### DISCUSSION THREAD

**Discussion by shushanhf** [dotnet/fsharp#13640] [General]
Is the fsharp project cross-platform  when porting is to a new CPU where the runtime had supported the LoongArch64 architecture ?

Should I have to port fsharp to a new architecture ?

**Comment by En3Tho** [comment]
F# is not different from C# in this perspective - F# compiler emits IL so it's supported anywhere IL is supported. I believe your work on runtime side should already enable F# code running on LA64.

There are alternative compilers like Fable (it supports compling F# to TypeScript, Python,  Rust, maybe more) but those will use different runtimes obviously

Is there any problem running simple Hello world F# app on LA64 for example?

**Reply by shushanhf** [reply]
> F# is not different from C# in this perspective - F# compiler emits IL so it's supported anywhere IL is supported. I believe your work on runtime side should already enable F# code running on LA64.
> 
> There are alternative compilers like Fable (it supports compling F# to TypeScript, Python, Rust, maybe more) but those will use different runtimes obviously
> 

Thanks very much !

> Is there any problem running simple Hello world F# app on LA64 for example?

```
qiao@Lap5K:~/fsharp/FSharpSample$ ~/work_qiao/dotnet-runtime-loongarch/.dotnet/dotnet build
用于 .NET 的 Microsoft (R) 生成引擎版本 17.0.0+c9eb9dd64
版权所有(C) Microsoft Corporation。保留所有权利。

  Determining projects to restore...
  All projects are up-to-date for restore.
FSC : error FS0193: 内部错误: Exception of type 'System.OutOfMemoryException' was thrown. [/home/qiao/fsharp/FSharpSample/src/Library/Library.fsproj]

生成失败。

FSC : error FS0193: 内部错误: Exception of type 'System.OutOfMemoryException' was thrown. [/home/qiao/fsharp/FSharpSample/src/Library/Library.fsproj]
    0 个警告
    1 个错误

已用时间 00:00:09.77
```

```
qiao@Lap5K:~/fsharp/FSharpSample/src/App$ ~/work_qiao/dotnet-runtime-loongarch/.dotnet/dotnet run Hello World
FSC : error FS0193: 内部错误: Exception of type 'System.OutOfMemoryException' was thrown. [/home/qiao/fsharp/FSharpSample/src/Library/Library.fsproj]

生成失败。请修复生成错误并重新运行。
```


<pre>
qiao@Lap5K:~/work_qiao/dotnet-runtime-loongarch$ file .dotnet/sdk/6.0.107/FSharp/*.dll
.dotnet/sdk/6.0.107/FSharp/fsc.dll:                                  PE32 executable (console) Intel 80386 Mono/.Net assembly, for MS Windows
.dotnet/sdk/6.0.107/FSharp/FSharp.Build.dll:                         PE32 executable (DLL) (console) Intel 80386 Mono/.Net assembly, for MS Windows
.dotnet/sdk/6.0.107/FSharp/FSharp.Compiler.Interactive.Settings.dll: PE32 executable (DLL) (console) Intel 80386 Mono/.Net assembly, for MS Windows
.dotnet/sdk/6.0.107/FSharp/FSharp.Compiler.Service.dll:              PE32 executable (DLL) (console) Intel 80386 Mono/.Net assembly, for MS Windows
.dotnet/sdk/6.0.107/FSharp/FSharp.Core.dll:                          PE32 executable (DLL) (console) Intel 80386 Mono/.Net assembly, for MS Windows
.dotnet/sdk/6.0.107/FSharp/FSharp.DependencyManager.Nuget.dll:       PE32 executable (DLL) (console) Intel 80386 Mono/.Net assembly, for MS Windows
.dotnet/sdk/6.0.107/FSharp/fsi.dll:                                  PE32 executable (console) Intel 80386 Mono/.Net assembly, for MS Windows
.dotnet/sdk/6.0.107/FSharp/Microsoft.Build.Framework.dll:            PE32 executable (DLL) (console) Intel 80386 Mono/.Net assembly, for MS Windows
.dotnet/sdk/6.0.107/FSharp/Microsoft.Build.Tasks.Core.dll:           PE32 executable (DLL) (console) Intel 80386 Mono/.Net assembly, for MS Windows
.dotnet/sdk/6.0.107/FSharp/Microsoft.Build.Utilities.Core.dll:       PE32 executable (DLL) (console) Intel 80386 Mono/.Net assembly, for MS Windows
.dotnet/sdk/6.0.107/FSharp/Microsoft.NET.StringTools.dll:            PE32 executable (DLL) (console) Intel 80386 Mono/.Net assembly, for MS Windows
.dotnet/sdk/6.0.107/FSharp/System.Resources.Extensions.dll:          PE32 executable (DLL) (console) Intel 80386 Mono/.Net assembly, for MS Windows
</pre>

**Reply by shushanhf** [reply]
I think my `fsc.dll` is not right.
How to build it when building the installer to generate the SDK.


The C# and il code are both OK.

**Reply by En3Tho** [reply]
I believe fsc.dll comes as part of dotnet official sdk. 
@vzarytovskii can you help here maybe?

**Reply by shushanhf** [reply]
The `fsc.dll` is PE32+ within the offical sdk6.0.7 but I want `fsc.dll` is PE32, not the AOT version.

The SDK is built by ourself while building the C# and IL is ok.
The Fsharp is built on x64-linux while disable the crossgen.

**Reply by En3Tho** [reply]
Pinging @dsyme too just in case

**Reply by vzarytovskii** [reply]
I think we only build anycpu and arm64 versions, but it should work on any platform dotnet works. OOMs can happen for different reasons (maybe tail prefixes cause it), we will need to investigate. 
Could you please create an issue for that?

Also, looping @KevinRansom in.

**Reply by shushanhf** [reply]
Thanks, I had created an issuse #13644
